std/string.d

```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
+                return false;
+
+            /* Since trailing spaces are removed,
+             * look ahead for anything that is not a trailing space
+             */
+            static if (isSomeString!Range)
             {
-                if (!changes)
-                    change();
-
-                ptrdiff_t j = result.length - nspaces;
-                auto ntabs = (((column - nspaces) % tabSize) + nspaces) / tabSize;
-                result.length = j + ntabs;
-                result[j .. j + ntabs] = '\t';
-                nwhite += ntabs - nspaces;
-                nspaces = 0;
+                foreach (c; _input)
+                {
+                    if (c != ' ' && c != '\t')
+                        return false;
+                }
+                return true;
             }
-            column = (column + tabSize) / tabSize * tabSize;
-            break;
-
-        case '\r':
-        case '\n':
-        case paraSep:
-        case lineSep:
-            // Truncate any trailing spaces or tabs
-            if (nwhite)
+            else
             {
-                if (!changes)
-                    change();
-                result = result[0 .. result.length - nwhite];
+                if (_input.empty)
+                    return true;
+                C c = _input.front;
+                if (c != ' ' && c != '\t')
+                    return false;
+                auto t = _input.save;
+                t.popFront();
+                foreach (c2; t)
+                {
+                    if (c2 != ' ' && c2 != '\t')
+                        return false;
+                }
+                return true;
             }
-            break;
+        }
 
-        default:
-            if (nspaces >= 2 && (column % tabSize) == 0)
+        @property C front()
+        {
+            //writefln("   front(): ntabs = %s nspaces = %s index = %s front = '%s'", ntabs, nspaces, index, getFront());
+            if (ntabs)
+                return '\t';
+            if (nspaces)
+                return ' ';
+            C c = getFront();
+            if (index)
+                return c;
+            dchar dc;
+            if (c < codeUnitLimit!(immutable(C)[]))
             {
-                if (!changes)
-                    change();
-
-                auto j = result.length - nspaces;
-                auto ntabs = (nspaces + tabSize - 1) / tabSize;
-                result.length = j + ntabs;
-                result[j .. j + ntabs] = '\t';
-                nwhite += ntabs - nspaces;
-                nspaces = 0;
-            }
-            if (c == ' ')
-            {   nwhite++;
-                nspaces++;
+                index = 1;
+                dc = c;
+                if (c == ' ' || c == '\t')
+                {
+                    // Consume input until a non-blank is encountered
+                    size_t startcol = column;
+                    C cx;
+                    static if (isSomeString!Range)
```

Comment:
Wouldn't it be better to work on [`_input.representation`](http://dlang.org/library/std/string/representation.html) rather than duplicating all this code for narrow strings?